polygeist/lib/Dialect/Polygeist/Transforms/LoopInternalization.cpp

```diff
@@ -349,99 +361,157 @@ struct LoopInternalization
     : public polygeist::impl::LoopInternalizationBase<LoopInternalization> {
   using LoopInternalizationBase<LoopInternalization>::LoopInternalizationBase;
 
-  void runOnOperation() final {
-    assert(CollectReadOnlyAccesses &&
-           "Limitation: only able to handle read only accesses currently");
+  void runOnOperation() final;
 
-    Operation *module = getOperation();
-    ModuleAnalysisManager mam(module, /*passInstrumentor=*/nullptr);
-    AnalysisManager am = mam;
-    auto &memAccessAnalysis =
-        am.getAnalysis<MemoryAccessAnalysis>().initialize(relaxedAliasing);
+private:
+  /// Construct a map from memref accesses in \p loop to their ideal memory
+  /// space.
+  void selectMemorySpace(LoopLikeOpInterface loop,
+                         const MemoryAccessAnalysis &memAccessAnalysis,
+                         DataFlowSolver &solver);
+
+  /// Transform a candidate loop.
+  template <typename T>
+  void transform(T loop, const MemoryAccessAnalysis &memAccessAnalysis,
+                 DataFlowSolver &solver) const;
 
-    auto walkResult = module->walk([&](FunctionOpInterface func) {
-      DataFlowSolver solver;
-      solver.load<dataflow::DeadCodeAnalysis>();
-      solver.load<dataflow::IntegerRangeAnalysis>();
-      if (failed(solver.initializeAndRun(func)))
-        return WalkResult::interrupt();
+private:
+  /// A map from a candidate loop to memref values used in the loop.
+  DenseMap<LoopLikeOpInterface, SmallVector<Value>> loopToMemref;
 
-      if (!isCandidate(func))
-        return WalkResult::advance();
+  /// Map from a candidate memref value to its ideal memory space.
+  DenseMap<Value, MemorySelector::MemorySpace> memrefToMemorySpace;
+};
+
+void LoopInternalization::runOnOperation() {
+  assert(CollectReadOnlyAccesses &&
```

Comment:
Can we make this option and `LoopInternalizationTileSize` proper pass options as part of the refactoring?

Reply:
I made the tile size a proper option and removed the other one (added a TODO).